demo/src/core/router.js

```diff
@@ -149,20 +194,47 @@ class Router extends EventTarget {
     return paramsKeys.every(k => this.#currentQueryParams[k] === params[k]);
   }
 
-  #updateCurrentRoute(route, queryParams) {
+  /**
+   * Updates the current route and dispatches the 'routechanged' event.
+   *
+   * @private
+   * @param {object} route - The route object.
+   * @param {object} queryParams - The query parameters associated with the route.
+   * @param {boolean} [popstate=false] - (Optional) if a popstate is at the origin of this route change.
+   */
+  #updateCurrentRoute(route, queryParams, popstate = false) {
     route.destroy();
     this.#currentRoute = route;
     this.#currentQueryParams = queryParams;
     route.start(queryParams);
-    this.dispatchEvent(new Event('routechanged'));
+    this.dispatchEvent(new CustomEvent('routechanged', {
+      detail: {
+        route,
+        popstate,
+        queryParams
+      }
+    }));
+  }
+
+  /**
+   * Initiates the router based on the current window location.
+   */
+  start() {
+    this.#initBase();
+
+    const url = new URL(window.location.href);
+    const path = url.pathname;
+    const queryParams = Object.fromEntries(url.searchParams.entries());
+
+    this.#handleRouteChange(path, queryParams);
   }
 
   /**
    * Initializes the base path for the router based on the current window location pathname.
    * If the current pathname matches a route, the base is set to the pathname with the last path segment removed.
    * If there is no matching route, the base is set to the full pathname.
    */
-  initBase() {
+  #initBase() {
     const pathname = window.location.pathname;
 
     this.base = this.findRoute(pathname) ?
```

Comment:
Since we're in the `start` function, do we really need an `initBase` function ?

What about a single function?
```javascript
  start() {
    const url = new URL(window.location.href);
    const path = url.pathname;
    const queryParams = Object.fromEntries(url.searchParams.entries());

    this.base = this.findRoute(path) ?
      path.replace(/\/[^/]+\/?$/, '/') :
      path;

    this.#handleRouteChange(path, queryParams);
  }
```